src/search/scaffold_search.rs

```diff
@@ -1,12 +1,20 @@
+use crate::search::structure_matching::substructure_match_fp;
+use bitvec::prelude::{BitSlice, BitVec};
 use rdkit::{substruct_match, ROMol, SubstructMatchParameters};
 use std::iter::Iterator;
 use std::sync::Arc;
 use std::sync::Mutex;
 
 const SCAFFOLDS: &str = include_str!("../../assets/standardized_scaffolds_20240405.json");
 
+pub struct Scaffold {
+    pub fp: Arc<Mutex<BitVec<u8>>>,
+    pub mol: Arc<Mutex<ROMol>>,
```

Comment:
Is this our first time doing things with `Arc` in the codebase? This is a relatively advanced type, do you know why we need it?

Reply:
It has something to do with not being able to pass ROMol's safely to threads (something about cxx voids). That said, I now realize we don't need an Arc<Mutex for the fingerprint...

Reply:
where does the ROMol have to cross thread boundaries?

Reply:
I guess the question is, how big is this vector? Will it introduce a lot of contention between threads? A lock is nice if you expect to _write_ data back. But if the vec is essentially read-only for the life of the program then we should consider another locking primitive. A mutex has _exclusive_ read and write. Something like a `RWLock` allows multiple readers and the writer has to wait for all readers to go away -- but if we never need a writer then maybe it works better?

Reply:
So could be that I'm missing something, but `ROMol`s don't seem to be playing nice with `Arc<RwLock`. Here's the error:

```
error[E0277]: `*const cxx::void` cannot be sent between threads safely
   --> src/search/scaffold_search.rs:16:1
    |
16  | / lazy_static::lazy_static! {
17  | |     pub static ref PARSED_SCAFFOLDS: Vec<Scaffold> = SCAFFOLDS
18  | |     .lines()
19  | |     .map(|l| serde_json::from_str::<serde_json::Value>(l).unwrap())
...   |
36  | |     .collect();
37  | | }
    | |_^ `*const cxx::void` cannot be sent between threads safely
    |
    = help: within `rdkit_sys::bridge::ro_mol::ffi::ROMol`, the trait `Send` is not implemented for `*const cxx::void`, which is required by `std::vec::Vec<Scaffold>: Sync`
    = note: required because it appears within the type `[*const cxx::void; 0]`
note: required because it appears within the type `cxx::opaque::Opaque`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cxx-1.0.124/src/opaque.rs:14:12
    |
14  | pub struct Opaque {
    |            ^^^^^^
note: required because it appears within the type `rdkit_sys::bridge::ro_mol::ffi::ROMol`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rdkit-sys-0.4.10/src/bridge/ro_mol.rs:20:18
    |
20  |         pub type ROMol;
    |                  ^^^^^
    = note: required for `cxx::shared_ptr::SharedPtr<rdkit_sys::bridge::ro_mol::ffi::ROMol>` to implement `Sync`
note: required because it appears within the type `rdkit::ROMol`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rdkit-0.4.11/src/graphmol/ro_mol.rs:8:12
    |
8   | pub struct ROMol {
    |            ^^^^^
    = note: required for `std::sync::RwLock<rdkit::ROMol>` to implement `Sync`
    = note: 1 redundant requirement hidden
    = note: required for `Arc<std::sync::RwLock<rdkit::ROMol>>` to implement `Sync`
note: required because it appears within the type `Scaffold`
   --> src/search/scaffold_search.rs:10:12
    |
10  | pub struct Scaffold {
    |            ^^^^^^^^
    = note: required for `Unique<Scaffold>` to implement `Sync`
note: required because it appears within the type `alloc::raw_vec::RawVec<Scaffold>`
   --> /Users/javierpineda/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:69:19
    |
69  | pub(crate) struct RawVec<T, A: Allocator = Global> {
    |                   ^^^^^^
note: required because it appears within the type `std::vec::Vec<Scaffold>`
   --> /Users/javierpineda/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:398:12
    |
398 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    |            ^^^
note: required by a bound in `Lazy`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.5.0/src/inline_lazy.rs:19:20
    |
19  | pub struct Lazy<T: Sync>(Cell<MaybeUninit<T>>, Once);
    |                    ^^^^ required by this bound in `Lazy`
    = note: this error originates in the macro `__lazy_static_create` which comes from the expansion of the macro `lazy_static::lazy_static` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `*const cxx::void` cannot be shared between threads safely
   --> src/search/scaffold_search.rs:16:1
    |
16  | / lazy_static::lazy_static! {
17  | |     pub static ref PARSED_SCAFFOLDS: Vec<Scaffold> = SCAFFOLDS
18  | |     .lines()
19  | |     .map(|l| serde_json::from_str::<serde_json::Value>(l).unwrap())
...   |
36  | |     .collect();
37  | | }
    | |_^ `*const cxx::void` cannot be shared between threads safely
    |
    = help: within `rdkit_sys::bridge::ro_mol::ffi::ROMol`, the trait `Sync` is not implemented for `*const cxx::void`, which is required by `std::vec::Vec<Scaffold>: Sync`
    = note: required because it appears within the type `[*const cxx::void; 0]`
note: required because it appears within the type `cxx::opaque::Opaque`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cxx-1.0.124/src/opaque.rs:14:12
    |
14  | pub struct Opaque {
    |            ^^^^^^
note: required because it appears within the type `rdkit_sys::bridge::ro_mol::ffi::ROMol`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rdkit-sys-0.4.10/src/bridge/ro_mol.rs:20:18
    |
20  |         pub type ROMol;
    |                  ^^^^^
    = note: required for `cxx::shared_ptr::SharedPtr<rdkit_sys::bridge::ro_mol::ffi::ROMol>` to implement `Sync`
note: required because it appears within the type `rdkit::ROMol`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rdkit-0.4.11/src/graphmol/ro_mol.rs:8:12
    |
8   | pub struct ROMol {
    |            ^^^^^
    = note: required for `std::sync::RwLock<rdkit::ROMol>` to implement `Sync`
    = note: 1 redundant requirement hidden
    = note: required for `Arc<std::sync::RwLock<rdkit::ROMol>>` to implement `Sync`
note: required because it appears within the type `Scaffold`
   --> src/search/scaffold_search.rs:10:12
    |
10  | pub struct Scaffold {
    |            ^^^^^^^^
    = note: required for `Unique<Scaffold>` to implement `Sync`
note: required because it appears within the type `alloc::raw_vec::RawVec<Scaffold>`
   --> /Users/javierpineda/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:69:19
    |
69  | pub(crate) struct RawVec<T, A: Allocator = Global> {
    |                   ^^^^^^
note: required because it appears within the type `std::vec::Vec<Scaffold>`
   --> /Users/javierpineda/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:398:12
    |
398 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    |            ^^^
note: required by a bound in `Lazy`
   --> /Users/javierpineda/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.5.0/src/inline_lazy.rs:19:20
    |
19  | pub struct Lazy<T: Sync>(Cell<MaybeUninit<T>>, Once);
    |                    ^^^^ required by this bound in `Lazy`
    = note: this error originates in the macro `__lazy_static_create` which comes from the expansion of the macro `lazy_static::lazy_static` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `get` exists for struct `Lazy<Vec<Scaffold>>`, but its trait bounds were not satisfied
   --> src/search/scaffold_search.rs:16:1
    |
16  | / lazy_static::lazy_static! {
17  | |     pub static ref PARSED_SCAFFOLDS: Vec<Scaffold> = SCAFFOLDS
18  | |     .lines()
19  | |     .map(|l| serde_json::from_str::<serde_json::Value>(l).unwrap())
...   |
36  | |     .collect();
37  | | }
    | |_^ method cannot be called on `Lazy<Vec<Scaffold>>` due to unsatisfied trait bounds
    |
   ::: /Users/javierpineda/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:398:1
    |
398 |   pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    |   ------------------------------------------------------------------------------------------------ doesn't satisfy `std::vec::Vec<Scaffold>: Sync`
    |
    = note: the following trait bounds were not satisfied:
            `*const cxx::void: Send`
            which is required by `std::vec::Vec<Scaffold>: Sync`
            `*const cxx::void: Sync`
            which is required by `std::vec::Vec<Scaffold>: Sync`
    = note: this error originates in the macro `__lazy_static_internal` which comes from the expansion of the macro `lazy_static::lazy_static` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
warning: `cheminee` (lib) generated 1 warning
error: could not compile `cheminee` (lib) due to 3 previous errors; 1 warning emitted

```

Reply:
That error is sufficiently nasty that I'm satisfied we can fix it later and/or when it's an actual source of problems.